### PR TITLE
Somewhat hacky but effective way to implement IDX/IMG LBA Seek.

### PR DESCRIPTION
**Pull request by Topaz-Reality**
Does what the title says. Seeks out the LBA Offset of the IDX and IMG from the ISO.
This is hacky, but as far as I have seen, it works with everything tested.

Addresses #348.

Tested ISOs:
-> KH2US
-> KH2FM
-> KH2CCZ

Of course, more testing is necessary. But this will do until we have a proper ISO library.

### CODE REVIEW

File: OpenKh.Command.IdxImg/Program.cs [hunk 1]
```diff
@@ -213,6 +226,49 @@ protected int OnExecute(CommandLineApplication app)
...
+                    isoStream.Position = (uint)(0x105 * 0x800);
+                    byte[] haystack = isoStream.ReadBytes(0x500);
+
+                    using (BinaryReader reader = new BinaryReader(new MemoryStream(haystack)))
+                    {
+                        for (int i = 0; i < 0x500; i++)
+                        {
+                            reader.BaseStream.Position = i;
+
+                            byte[] hayRead = reader.ReadBytes(0x0B);
```
Comment (Xeeynamo) [inline OpenKh.Command.IdxImg/Program.cs:245]: Suggestion here to replace `0x0B` with a constant (eg. `NeedleLength`) and use it when declaring those arrays above (eg. `imgNeedle = new byte[NeedleLength]` so the reason why it is required to read that amount of byte is more explicit.

File: OpenKh.Command.IdxImg/Program.cs [hunk 2]
```diff
@@ -213,6 +226,49 @@ protected int OnExecute(CommandLineApplication app)
...
+
+                if (IdxIsoBlock == -1 || ImgIsoBlock == -1)
+                {
+                    byte[] imgNeedle = new byte[] { 0x01, 0x09, 0x4B, 0x48, 0x32, 0x2E, 0x49, 0x4D, 0x47, 0x3B, 0x31 };
+                    byte[] idxNeedle = new byte[] { 0x01, 0x09, 0x4B, 0x48, 0x32, 0x2E, 0x49, 0x44, 0x58, 0x3B, 0x31 };
+
+                    isoStream.Position = (uint)(0x105 * 0x800);
+                    byte[] haystack = isoStream.ReadBytes(0x500);
+
+                    using (BinaryReader reader = new BinaryReader(new MemoryStream(haystack)))
```
Comment (Xeeynamo) [inline OpenKh.Command.IdxImg/Program.cs:239]: I believe it is not required to read a `0x500` block and using it in the `BinaryReader`. Can you do `new BinaryReader(isoStream)` instead? If you are concerned about performance, you can wrap `isoStream` with a `BufferedStream`.

File: OpenKh.Command.IdxImg/Program.cs [hunk 3]
```diff
@@ -199,11 +199,24 @@ private class InjectCommand
...
-            public long IdxIsoBlock { get; set; } = 1840;
+            [Option(CommandOptionType.SingleValue, Description = "ISO block for KH2.IDX. Determined automatically by default.", ShortName = "idx", LongName = "idx-offset")]
+            public int IdxIsoBlock { get; set; } = -1;
 
-            [Option(CommandOptionType.SingleValue, Description = "ISO block for KH2.IMG. By default is 671693 for KH2FM", ShortName = "img", LongName = "img-offset")]
-            public long ImgIsoBlock { get; set; } = 671693;
+            [Option(CommandOptionType.SingleValue, Description = "ISO block for KH2.IMG. Determined automatically by default", ShortName = "img", LongName = "img-offset")]
+            public int ImgIsoBlock { get; set; } = -1;
+
+            bool CompareArrays(byte[] input1, byte[] input2)
```
Comment (Xeeynamo) [inline OpenKh.Command.IdxImg/Program.cs:208]: No need to create a dedicated function. `input1.SequenceEqual(input2)` will be good enough and it's fast when comparing two arrays of the same length!

File: OpenKh.Command.IdxImg/Program.cs [hunk 4]
```diff
@@ -213,6 +226,49 @@ protected int OnExecute(CommandLineApplication app)
...
+
+                            if (imgCmp || idxCmp)
+                            {
+                                reader.BaseStream.Position -= 0x24;
+
+                                byte[] blockStack = reader.ReadBytes(4);
+                                byte[] blockCorrect = new byte[] { blockStack[3], blockStack[2], blockStack[1], blockStack[0] };
+
+                                if (idxCmp && IdxIsoBlock == -1)
+                                    IdxIsoBlock = BitConverter.ToInt32(blockCorrect);
```
Comment (Xeeynamo) [inline OpenKh.Command.IdxImg/Program.cs:258]: I never considered to use `BitConverter.ToInt32`. Thanks for sharing this tip :)

File: OpenKh.Command.IdxImg/Program.cs [hunk 5]
```diff
@@ -213,6 +226,49 @@ protected int OnExecute(CommandLineApplication app)
 
                 using var isoStream = File.Open(InputIso, FileMode.Open, FileAccess.ReadWrite);
 
+                // Begin being hacky.
```
Comment (Xeeynamo) [inline OpenKh.Command.IdxImg/Program.cs:229]: No need to specify this is an hack as long as it works ^^

**Review by Xeeynamo — CHANGES_REQUESTED** [review]
Small tip: use `var` instead of the explicit type when possible

**Comment by Xeeynamo** [comment]
Another suggestion: if you still plan to use that buffer of `0x500` you can actually get rid of `BinaryReader` and the `ReadBytes` all together. .NET Core 3.1 comes with `Span<T>`, so you can slice a portion of an array with no performance impact. For instance, those two expressions are functionally equivalent, but the second one is more performant:

```csharp
var reader = new BinaryReader(new MemoryStream(buffer));
reader.BaseStream.Position = 5;
var data = reader.ReadBytes(11)
```
with Span<T>:
```csharp
var data = new Span(buffer, 5, 11);
```

**Comment by Topaz-Reality** [comment]
Changes made.

**Review by Xeeynamo — APPROVED** [review]

**Pull request merged**